File: src/ProjectR.Scripting/content/scripts/afflictions/Debuffs/Silence.lua
function GetName()
  return "Silence"
end

function GetType()
  return Debuff
end

function GetHookPoints()
  return { TurnTriggered }
end

function OnAttachment(char)
  char:SetSilenced(true)
  char:SetVar("sil_Chance", 1500)
end

function OnRemoval(char)
  char:SetSilenced(false)
end

function Trigger(hookPoint, char)
  if hookPoint ~= TurnTriggered then
    return
  end

  local silChance = char:GetVar("sil_Chance")
  local silResi = GetStats(char):GetTotalStat(SIL) * 3

  if Roll(0, 99) > silChance - silResi then
    this:RemoveFrom(char)
  else
    char:SetVar("sil_Chance", silChance / 3)
  end
end

File: src/ProjectR.Scripting/content/scripts/afflictions/Debuffs/Silence.cs
using ProjectR.Interfaces.Helper;
using ProjectR.Interfaces.Model;
using ProjectR.Interfaces.Model.Stats;

namespace ProjectR.Scripting.Afflictions
{
    public class Silence : Affliction
    {
        public override string Name { get { return "Silence"; } }

        public override AfflictionType Type { get { return AfflictionType.Debuff; } }

        protected override HookPoint[] HookPoints
        {
            get
            {
                return new[]
                {
                    HookPoint.TurnTriggered, 
                };
            }
        }

        protected override void OnAttachment(ICharacter character)
        {
            character.IsSilenced = true;
            SetVar(character, "sil_Chance", 1500);
        }

        protected override void OnRemoval(ICharacter character)
        {
            character.IsSilenced = false;
        }

        protected override void OnTurnTriggered(ICharacter character)
        {
            var silChance = GetVar(character, "sil_Chance");
            var silResi = character.Stats.GetTotalStat(DebuffResistance.SIL) * 3;

            if (RHelper.Roll(0, 99) > (silChance - silResi))
            {
                RemoveFrom(character);
            }
            else
            {
                SetVar(character, "sil_Chance", silChance / 3);
            }
        }
    }
}